From 07afa8267805d77d25cba33f8e362968f54ad951 Mon Sep 17 00:00:00 2001
From: Li Jun <jun.li@nxp.com>
Date: Sat, 17 Jun 2017 01:54:55 +0800
Subject: [PATCH 1953/5242] MLK-15074-10 usb: gadget: utp: correct the set_fs
 condition

commit  18eb1cda1b61ee577e43fec7a0a7f6ed974fdb71 from
https://source.codeaurora.org/external/imx/linux-imx.git

set_fs() should be done in case:
1. CONFIG_FSL_UTP is not enabled.
2. CONFIG_FSL_UTP is enabled but is_utp_device is false.

Signed-off-by: Li Jun <jun.li@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/usb/gadget/function/f_mass_storage.c |   12 ++++++++++++
 1 file changed, 12 insertions(+)

diff --git a/drivers/usb/gadget/function/f_mass_storage.c b/drivers/usb/gadget/function/f_mass_storage.c
index 2d03e56..07efb6a 100644
--- a/drivers/usb/gadget/function/f_mass_storage.c
+++ b/drivers/usb/gadget/function/f_mass_storage.c
@@ -2494,6 +2494,18 @@ static int fsg_main_thread(void *common_)
 	/* Allow the thread to be frozen */
 	set_freezable();
 
+	/*
+	 * Arrange for userspace references to be interpreted as kernel
+	 * pointers.  That way we can pass a kernel pointer to a routine
+	 * that expects a __user pointer and it will work okay.
+	 */
+#ifdef CONFIG_FSL_UTP
+	if (!is_utp_device(common->fsg))
+		set_fs(get_ds());
+#else
+	set_fs(get_ds());
+#endif
+
 	/* The main loop */
 	while (common->state != FSG_STATE_TERMINATED) {
 		if (exception_in_progress(common) || signal_pending(current)) {
-- 
1.7.9.5

